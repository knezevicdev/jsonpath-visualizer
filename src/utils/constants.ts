export const ARRAY_GROUP_LIMIT = 100;
