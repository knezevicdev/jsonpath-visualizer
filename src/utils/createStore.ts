export type TJson = null | Record<string, unknown> | any[];

export const createStore = () => {
  return {
    json: null as TJson,
    jsonPath: '' as string,
  };
};

export type TStore = ReturnType<typeof createStore>;
