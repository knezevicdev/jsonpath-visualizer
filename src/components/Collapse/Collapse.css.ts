import styled, { css } from 'styled-components';

import { Wrapper as StyledWrapper } from 'styled/components';
import colors from 'styled/colors';
import { matchedStyle } from 'styled/helpers';

interface CollapseProps {
  opened: boolean;
}

export const Wrapper = styled(StyledWrapper)<CollapseProps>`
  ${({ opened }) =>
    !opened &&
    css`
      display: flex;
    `}
`;

export const Value = styled.span<CollapseProps>`
  ${({ opened }) =>
    !opened &&
    css`
      display: flex;
    `}
`;

export const Content = styled.div<CollapseProps>`
  ${({ opened }) =>
    !opened &&
    css`
      display: none;
    `}
`;

export const Bracket = styled.span`
  font-weight: bold;
  color: ${colors.white};

  [data-matched='true'] & {
    ${matchedStyle}
  }
`;

export const Ellipsis = styled.span`
  display: inline-block;
  color: ${colors.orange};
  font-size: 18px;
  line-height: 10px;
  cursor: pointer;
`;
