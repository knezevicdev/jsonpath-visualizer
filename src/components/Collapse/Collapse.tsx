import React, { FunctionComponent, useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { observer } from 'mobx-react';

import { useStore } from 'utils/store';
import { FieldName } from 'styled/components';
import { ExpandedIcon, CollapsedIcon } from 'components';

import { Wrapper, Value, Content, Bracket, Ellipsis } from './Collapse.css';

type CollapseProps = {
  opened: boolean;
  name: string;
  path: string;
  isArray?: boolean;
  renderContent: () => JSX.Element[];
};

const Collapse: FunctionComponent<CollapseProps> = ({ opened, name, path, isArray, renderContent }) => {
  const [isOpened, setIsOpened] = useState(opened);
  const alreadyOpened = useRef(false);
  const { [path]: matched } = useStore();

  useEffect(() => {
    if (opened) {
      alreadyOpened.current = true;
    }

    setIsOpened(opened);
  }, [opened]);

  const toggleOpened = () => {
    if (!isOpened) {
      alreadyOpened.current = true;
    }

    setIsOpened(!isOpened);
  };

  return (
    <Wrapper opened={isOpened}>
      <span onClick={toggleOpened}>
        {isOpened ? <ExpandedIcon /> : <CollapsedIcon />}&nbsp;
        {name && name.length && <FieldName>{name}:&nbsp;</FieldName>}
      </span>
      <Value opened={isOpened} data-matched={matched}>
        <Bracket>{isArray ? '[' : '{'}</Bracket>
        {(alreadyOpened.current || isOpened) && <Content opened={isOpened}>{renderContent()}</Content>}
        {!isOpened && (
          <div onClick={toggleOpened}>
            <Ellipsis>...</Ellipsis>
          </div>
        )}
        <Bracket>{isArray ? ']' : '}'}</Bracket>
      </Value>
    </Wrapper>
  );
};

Collapse.propTypes = {
  opened: PropTypes.bool.isRequired,
  name: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
  isArray: PropTypes.bool,
  renderContent: PropTypes.func.isRequired,
};

export default observer(Collapse);
