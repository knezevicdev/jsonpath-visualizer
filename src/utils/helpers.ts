import { transform as _transform, isObject as _isObject, mapKeys as _mapKeys, assign as _assign } from 'lodash';
import { JSONPath } from 'jsonpath-plus';

import { TJson } from './store';

// eslint-disable-next-line
export const toType = (obj: any): string => {
  let type = getType(obj);
  if (type === 'number') {
    if (isNaN(obj)) {
      type = 'nan';
    } else if ((obj | 0) != obj) {
      //bitwise OR produces integers
      type = 'float';
    } else {
      type = 'integer';
    }
  }
  return type;
};

//source: http://stackoverflow.com/questions/7390426/better-way-to-get-type-of-a-javascript-variable/7390612#7390612
const getType = (obj: any): string => {
  return {}.toString
    .call(obj)
    .match(/\s([a-zA-Z]+)/)[1]
    .toLowerCase();
};

// eslint-disable-next-line
export const objectPaths = (obj: any): Record<string, unknown> => {
  return _transform(
    obj,
    (result: Record<string, unknown>, value: unknown, key: string) => {
      if (_isObject(value)) {
        const flatMap = _mapKeys(objectPaths(value), function (_mvalue, mkey) {
          return `${key}.${mkey}`;
        });

        _assign(result, flatMap);
      } else {
        result[String(key)] = false;
      }

      return result;
    },
    {},
  );
};

export const matchPaths = (json: TJson, path: string): string[] => {
  const paths = JSONPath({
    path,
    json,
    resultType: 'pointer',
  }).map((pointer: string) => pointer.replace(/\//g, '.').substring(1));

  if (paths.includes('')) {
    return [''];
  }

  return paths.reduce((matchedPaths: string[], path: string) => {
    for (const matchedPath in matchedPaths) {
      if (path.includes(matchedPath)) {
        return matchedPaths;
      }
    }

    return [...matchedPaths, path];
  }, []);
};
