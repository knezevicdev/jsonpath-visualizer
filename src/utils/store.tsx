import React, { ComponentType } from 'react';
import { useLocalStore } from 'mobx-react';
import { reaction, comparer, keys, remove } from 'mobx';

import { matchPaths } from 'utils/helpers';

export type TJson = null | Record<string, unknown> | any[];

const storeContext = React.createContext<Record<string, any> | null>(null);

export const StoreProvider: React.FC = ({ children }) => {
  const store = useLocalStore<Record<string, any>>(() => ({
    json: null as TJson,
    jsonPath: '' as string,
    jsonUploadTime: 0 as number,
    matchedPaths: [] as string[],
  }));

  reaction(
    () => store.jsonPath,
    (jsonPath) => {
      const matchedPaths = matchPaths(store.json, jsonPath);

      store.matchedPaths.forEach((oldMatchedPath: string) => {
        if (!matchedPaths.includes(oldMatchedPath)) {
          store[oldMatchedPath] = false;
        }
      });

      matchedPaths.forEach((matchedPath: string) => {
        if (!store[matchedPath]) {
          store[matchedPath] = true;
        }
      });

      store.matchedPaths = matchedPaths;
    },
  );

  reaction(
    () => store.json,
    () => {
      keys(store).forEach((key: string) => {
        if (!['json', 'jsonPath', 'matchedPaths', 'jsonUploadTime'].includes(key)) {
          remove(store, key);
        }
      });

      store.matchedPaths = [];
      store.jsonPath = '';
    },
    { equals: comparer.shallow },
  );

  return <storeContext.Provider value={store}>{children}</storeContext.Provider>;
};

export const withStore = (WrappedComponent: ComponentType): React.FC => {
  const StoreHOC: React.FC = (props) => {
    return (
      <StoreProvider>
        <WrappedComponent {...props} />
      </StoreProvider>
    );
  };

  return StoreHOC;
};

export const useStore = (): Record<string, any> => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
