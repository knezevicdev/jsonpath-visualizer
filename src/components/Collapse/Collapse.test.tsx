import React from 'react';
import { shallow } from 'enzyme';

import { StoreProvider } from 'utils/store';

import Collapse from './Collapse';

describe('Collapse', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(
      <StoreProvider>
        <Collapse opened={true} name="" path="" renderContent={() => []} />
      </StoreProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
