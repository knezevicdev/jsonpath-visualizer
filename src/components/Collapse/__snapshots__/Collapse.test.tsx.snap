// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse renders snapshots 1`] = `
<ContextProvider
  value={
    Object {
      "json": null,
      "jsonPath": "",
      "jsonUploadTime": 0,
      "matchedPaths": Array [],
    }
  }
>
  <Collapse
    name=""
    opened={true}
    path=""
    renderContent={[Function]}
  />
</ContextProvider>
`;
